src/track/globaltrackcache.cpp

```diff
@@ -300,296 +422,309 @@ TrackPointer GlobalTrackCache::lookupInternal(
+        // Insert item by track location
+        DEBUG_ASSERT(m_tracksByCanonicalLocation.find(
+                trackRef.getCanonicalLocation()) == m_tracksByCanonicalLocation.end());
+        m_tracksByCanonicalLocation.insert(std::make_pair(
                 trackRef.getCanonicalLocation(),
-                item);
-        Stat::track(kInsertByCanonicalLocationCounter, Stat::COUNTER, kStatCounterFlags, 1);
+                plainPtr));
     }
-    return GlobalTrackCacheResolver(
-            std::move(cacheResolver),
+    pCacheResolver->initLookupResult(
             GlobalTrackCacheLookupResult::MISS,
-            std::move(trackRef),
-            std::move(pTrack));
+            std::move(strongPtr),
+            std::move(trackRef));
 }
 
-TrackRef GlobalTrackCache::updateTrackIdInternal(
-        const TrackPointer& pTrack,
-        const TrackRef& trackRef,
+TrackRef GlobalTrackCache::initTrackId(
+        const TrackPointer& strongPtr,
+        TrackRef trackRef,
         TrackId trackId) {
+    DEBUG_ASSERT(strongPtr);
+    DEBUG_ASSERT(!trackRef.getId().isValid());
     DEBUG_ASSERT(trackId.isValid());
-    if (trackRef.getId() != trackId) {
-        DEBUG_ASSERT(m_tracksById.end() == m_tracksById.find(trackId));
-        TrackRef trackRefWithId(trackRef, trackId);
-        Item item(trackRefWithId, pTrack);
-        m_tracksById.insert(
-                item.ref.getId(),
-                item);
-        m_tracksByCanonicalLocation.insert(
-                item.ref.getCanonicalLocation(),
-                item);
-        return trackRefWithId;
-    }
-    return trackRef;
-}
-
-void GlobalTrackCache::evict(
-        Track* pTrack) {
-    DEBUG_ASSERT(pTrack != nullptr);
-    const auto trackRef = TrackRef::fromFileInfo(
-            pTrack->m_fileInfo,
-            pTrack->m_record.getId());
-    GlobalTrackCacheLocker cacheLocker;
-    Track* pEvictedTrack = evictInternal(&cacheLocker, trackRef);
-    // The cache might have been unlocked during the callback!
-    DEBUG_ASSERT((nullptr == pEvictedTrack) || (pEvictedTrack == pTrack));
+    DEBUG_ASSERT(m_tracksByCanonicalLocation.find(
+            trackRef.getCanonicalLocation()) != m_tracksByCanonicalLocation.end());
+    TrackRef trackRefWithId(trackRef, trackId);
+
+    // Insert item by id
+    DEBUG_ASSERT(m_tracksById.find(trackId) == m_tracksById.end());
+    m_tracksById.insert(std::make_pair(
+            trackId,
+            strongPtr.get()));
+
+    strongPtr->initId(trackId);
+    DEBUG_ASSERT(createTrackRef(*strongPtr) == trackRefWithId);
+
+    return trackRefWithId;
 }
 
-GlobalTrackCache::Item GlobalTrackCache::purgeInternal(
-        const TrackRef& trackRef) {
-    if (kLogger.debugEnabled()) {
-        kLogger.debug()
-                << "Purging track"
-                << trackRef;
-    }
+bool GlobalTrackCache::evictAndDelete(
+        Track* plainPtr) {
+    DEBUG_ASSERT(plainPtr);
+    GlobalTrackCacheLocker cacheLocker;
```

Comment:
Before the lock here, bad things can happen, because the shared_ptr is already gone.

Reply:
OK, I got it. It should work.
I think there are some misleading comments and some confusing code parts. 
I will try to improve it based on this concept.